## analysis the cuffdiff
path="/data1/linhua/QIANLAB/PROJECT/Long-Noncoding-RNA-project/TOPHAT_MAPPING_TEST/DIFF_ANALYSIS/"

setwd(path)

dir()


cat(readLines("Again_stranded_cuff_diff.sh"),sep="\n")

library(cummeRbund)

cuff<-readCufflinks("Again_stranded_diffout-2016-10-25/")

cuff

#Several plotting methods are available that allow for quality-control or global analysis of cu✏inks data.
# 1.A good place to begin is to evaluate the quality of the model fitting.

disp<-dispersionPlot(genes(cuff))
disp

genes.scv<-fpkmSCVPlot(genes(cuff))
isoforms.scv<-fpkmSCVPlot(isoforms(cuff))


# 2.To assess the distributions of FPKM scores across samples, you can use the csDensity plot (Figure 1).
dens<-csDensity(genes(cuff))
dens
densRep<-csDensity(genes(cuff),replicates=T)
densRep

# 3.Boxplots can be visualized using the csBoxplot method (Figure 2).
b<-csBoxplot(genes(cuff))
b
brep<-csBoxplot(genes(cuff),replicates=T)
brep

# 4.A matrix of pairwise scatterplots can be drawn using the csScatterMatrix() method.

s<-csScatterMatrix(genes(cuff))
s

dend<-csDendro(genes(cuff))
#dend.rep<-csDendro(genes(cuff),replicates=T)


# 5.MvsA plots can be useful to determine any systematic bias that may be present between conditions.
m<-MAplot(genes(cuff),"1","2")

# 6.Volcano plots are also available for the Cu↵Data objects.
v<-csVolcanoMatrix(genes(cuff))
v

## Meta info retrive
# Run-level information such as run parameters, and sample information can be accessed from a Cu↵Set object by using the runInfo and replicates methods:
runInfo(cuff)
replicates(cuff)

# Condition and feature names
# Vectors of sample names and feature names are available by using the samples and featureNames methods

sample.names<-samples(genes(cuff))
gene.featurenames<-featureNames(genes(cuff))
gene.features<-annotation(genes(cuff))
gene.fpkm<-fpkm(genes(cuff))


# Convenience functions
# To facilitate Bioconductor-like operations, an ’FPKM-matrix’ can be returned easily using the fpkmMatrix method

gene.matrix<-fpkmMatrix(genes(cuff))

#A matrix of replicate FPKM values can be retrieved by using repFpkmMatrix

gene.rep.matrix<-repFpkmMatrix(genes(cuff))

# Similarly, a matrix of normalized counts can be generated by using countMatrix

gene.count.matrix<-countMatrix(genes(cuff))

#
gene.repFpkm<-repFpkm(genes(cuff))

gene.counts<-count(genes(cuff))

isoform.fpkm<-fpkm(isoforms(cuff))

gene.diff<-diffData(genes(cuff))

# specific genes
test_gene<-head(gene.features$gene_id,n=40)
test<-getGenes(cuff,test_gene)
myGenes<-test
h<-csHeatmap(myGenes,cluster="both")
h
b<-expressionBarplot(myGenes)
b

# gene diff test
gene.diff<-diffData(genes(cuff))
head(gene.diff)

test_s<-subset(gene.diff,gene.diff$q_value<0.05 & abs(gene.diff$log2_fold_change)>1 )

write.csv(gene.diff,file = "Wang_Yue_RNA-Seq_all_genes.csv",row.names = F)

write.csv(test_s,file = "Wang_Yue_RNA-Seq.csv",row.names = F)
nrow(test_s)

MyGene<-getGenes(cuff,test_s$gene_id)
h<-csHeatmap(myGenes,cluster="both",border = T)
h










##################@@@@@@@@@@@@@@@@@@@@@@@@@@@@@%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
selected_gene_diff<-subset(gene.diff,gene.diff$q_value<0.01 & gene.diff$log2_fold_change>1 & gene.diff$significant=="yes")
nrow(selected_gene_diff)
head(selected_gene_diff)

test<-getGenes(cuff,selected_gene_diff$gene_id)


library(rtracklayer)
TAIR_TE_gene<-import.bed("TAIR10_TE_gene.bed")
TAIR_TE_gene
head(TAIR_TE_gene)
TAIR_TE_gene@elementMetadata$name
ID_TAIR_TE_gene<-substr(TAIR_TE_gene@elementMetadata$name,start = 1,stop = 9)
ID_TAIR_TE_gene
nchar(TAIR_TE_gene@elementMetadata$name)
unique(nchar(TAIR_TE_gene@elementMetadata$name))
write(ID_TAIR_TE_gene,file = "ID_TAIR_TE_gene.txt")


write(ID_TAIR_TE_gene,file = "ID_TAIR_TE_gene.txt")

TAIR_TE<-import.bed("TAIR10_TE.bed")

TAIR_TE@elementMetadata@listData$name

###### 
#@: http://blog.sina.com.cn/s/blog_6caea8bf010159dt.html


overlaps<-intersect(ID_TAIR_TE_gene,selected_gene_diff$gene_id)






